pfunk/tests/_tests.py

```diff
@@ -23,91 +23,95 @@ def tests():
     return sorted(_tests.keys())
 
 
-def run(name):
+def run(name, database):
     """ Runs a selected test. """
-    _tests[name].run()
+    _tests[name].run(database)
 
 
 def plot(name, show=False):
     """ Generates a selected test's plots. """
     _tests[name].plot(show)
 
 
-def analyse(name):
+def analyse(name, database):
     """ Analyse results for a selected test. """
-    return _tests[name].analyse()
+    return _tests[name].analyse(database)
+
+
+def results_writer_generator(name, date, file):
```

Comment:
Why this instead of just having them create an instance themselves?

Reply:
To avoid coupling the tests to the format in which we store test results. This is an example of the Dependency Inversion Principle.

When I made the database objects, the tests were directly creating instances of the file writers. Introducing this late binding let me switch between the two and compare results. If/when we want to add other formats, like JUnit-compatible XML or an RDBMS, we pass different callables instead of editing the test result class.